k8s/azure/aks-cluster.tf

```diff
@@ -0,0 +1,104 @@
+# FIXME: For real deployment we should store the terraform state
+# in cloud storage rather than just the current directory, terraform
+# supports Azure blob storage directly. This means configuration
+# doesn't need to be on a single machine somewhere.
+#
+# See https://www.terraform.io/language/settings/backends/azurerm
+#
+#terraform {
+#  backend "azurerm" {
+#    resource_group_name  = "kernelci-tf-storage"
+#    storage_account_name = "kernelci-tf"
+#    container_name       = "tfstate"
+#    key                  = "workers.terraform.tfstate"
+#  }
+#}
+
+provider "azurerm" {
+  features {}
+}
+
+# We assign all clusters to the same resource group, this is purely for
+# accounting purposes so it doesn't matter where the resource group is
+resource "azurerm_resource_group" "workers" {
+  name     = "kernelci-workers"
+  location = "East US"
+
+  tags = {
+    environment = "kernelci-workers"
+  }
+}
+
+locals {
+  zones = toset([
+    "uksouth",
+    "eastus",
+  ])
+}
+
+resource "azurerm_kubernetes_cluster" "workers" {
+  for_each = local.zones
+
+  name                = "${each.key}-workers-aks"
+  location            = each.key
+  resource_group_name = azurerm_resource_group.workers.name
+  dns_prefix          = "${each.key}-workers-k8s"
+
+  # Automatically roll out upgrades from AKS
+  automatic_channel_upgrade = "stable"
+
+  # Single always present node as AKS requires a default node pool -
+  # Terraform and/or AKS don't let us tag this as a spot instance and
+  # ideally we can scale the builders down to 0 so this is a small
+  # instance not tagged for work.
+  default_node_pool {
+    name            = "default"
+    node_count      = 1
+    vm_size         = "Standard_DS2_v2"
+    os_disk_size_gb = 30
+
+    node_labels = {
+      "kernelci/management" = "management"
+    }
+  }
+
+  service_principal {
+    client_id     = var.appId
+    client_secret = var.password
+  }
+
+  role_based_access_control {
+    enabled = true
+  }
+
+  tags = {
+    environment = "kernelci"
+  }
+}
+
+# The actual workers where submitted jobs will execute
+resource "azurerm_kubernetes_cluster_node_pool" "workers" {
+  for_each = azurerm_kubernetes_cluster.workers
+
+  name = "workers"
+  kubernetes_cluster_id = each.value.id
+
+  # FIXME: This is a very small node, what are we using?
```

Comment:
For the "normal" builders, our current clusters are 8-core.  Either `Standard_D8s_vX` (seems we have some v3, some v5 for recently created clusters.)
For the "big" builders, they're 32-core. `Standard_F32s_v2`

Reply:
Thanks. I'm wondering if we should either standardise on the 32 core instances for everything (and pack more jobs on there) or take a hit to the allmodconfig builds and standardise on 16 cores (though I think the pahole builds need the big machines so we probably need to keep 32 cores). We should also figure out if we need the big builders to be separate clusters or we can just have 2 nodegroups on the same cluster - the latter seems better since it would give the scheduler more flexibility, we can still use nodeSelectors on the jobs to force jobs onto one of the nodegroups.

This is one area where Karpenter makes life a whole lot easier than cluster-autoscaler.